presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -0,0 +1,575 @@
+
+    /**
+     * 5^13 fits in 2^31.
+     */
+    private static final int MAX_POWER_FIVE_INT = 13;
+    /**
+     * 5^x. All unsigned values.
+     */
+    private static final int[] POWER_FIVES_INT = new int[MAX_POWER_FIVE_INT + 1];
+
+    private static final Unsafe unsafe;
+
+    static {
+        try {
+            // fetch theUnsafe object
+            Field field = Unsafe.class.getDeclaredField("theUnsafe");
+            field.setAccessible(true);
+            unsafe = (Unsafe) field.get(null);
+            if (unsafe == null) {
+                throw new RuntimeException("Unsafe access not available");
+            }
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+
+        for (int i = 0; i < POWERS_OF_TEN.length; ++i) {
+            POWERS_OF_TEN[i] = unscaledDecimal(BigInteger.TEN.pow(i));
+        }
+
+        for (int i = 0; i < POWER_FIVES_INT.length; ++i) {
+            POWER_FIVES_INT[i] = BigInteger.valueOf(5).pow(i).intValue();
+        }
+    }
+
+    public static Slice unscaledDecimal()
+    {
+        return Slices.allocate(UNSCALED_DECIMAL_128_SLICE_LENGTH);
+    }
+
+    public static Slice unscaledDecimal(Slice decimal)
+    {
+        return Slices.copyOf(decimal);
+    }
+
+    public static Slice unscaledDecimal(String unscaledValue)
+    {
+        return unscaledDecimal(new BigInteger(unscaledValue));
+    }
+
+    public static Slice unscaledDecimal(BigInteger unscaledValue)
+    {
+        byte[] bytes = unscaledValue.abs().toByteArray();
+
+        if (bytes.length > UNSCALED_DECIMAL_128_SLICE_LENGTH
+                || (bytes.length == UNSCALED_DECIMAL_128_SLICE_LENGTH && (bytes[0] & SIGN_BYTE_MASK) != 0)) {
+            throwOverflowException();
+        }
+
+        // convert to little-endian order
+        reverse(bytes);
+
+        Slice decimal = Slices.allocate(UNSCALED_DECIMAL_128_SLICE_LENGTH);
+        decimal.setBytes(0, bytes);
+        if (unscaledValue.signum() < 0) {
+            setNegative(decimal, true);
+        }
+
+        throwIfOverflows(decimal);
+
+        return decimal;
+    }
+
+    public static Slice unscaledDecimal(long unscaledValue)
+    {
+        long[] longs = new long[NUMBER_OF_LONGS];
+        if (unscaledValue < 0) {
+            longs[0] = -unscaledValue;
+            longs[1] = SIGN_LONG_MASK;
+        }
+        else {
+            longs[0] = unscaledValue;
+        }
+        return Slices.wrappedLongArray(longs);
+    }
+
+    public static BigInteger unscaledDecimalToBigInteger(Slice decimal)
+    {
+        byte[] bytes = decimal.getBytes(0, UNSCALED_DECIMAL_128_SLICE_LENGTH);
+        // convert to big-endian order
+        reverse(bytes);
+        bytes[0] &= ~SIGN_BYTE_MASK;
+        return new BigInteger(isNegative(decimal) ? -1 : 1, bytes);
+    }
+
+    public static long unscaledDecimalToUnscaledLong(Slice decimal)
+    {
+        long lo = getLong(decimal, 0);
+        long hi = getLong(decimal, 1);
```

Comment:
I think it might be confusing with some kind of [low, high] range. I use `v0`, `vx` anyway, so maybe I will stick with `lo` and `hi` for now.